mrrc/storage.py

```diff
@@ -0,0 +1,235 @@
+"""
+Copyright (c) 2021 Red Hat, Inc
+All rights reserved.
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+import logging
+import os
+from typing import Callable, Dict, List
+
+from boto3 import session
+from botocore.errorfactory import ClientError
+
+from mrrc.utils.files import read_sha1
+
+logger = logging.getLogger(__name__)
+
+PRODUCT_META_KEY = "rh-products"
+CHECKSUM_META_KEY = "checksum"
+
+
+class S3Client(object):
+    """The S3Client is a wrapper of the original boto3 s3 client, which will provide
+       some convenient methods to be used in the mrrc uploader.
+    """
+
+    def __init__(self, extra_conf=None) -> None:
+        s3_session = session.Session()
+        self.client = s3_session.resource(
+            's3',
+            config=extra_conf,
```

Comment:
For testing we need to use a internal s3 compatible environment, which need to set it through "endpoint_url" here. It is not supported through aws hosted configuration file, so we need to add extra ways to pass it in.

Reply:
Yes. we need a "endpoint_url=xxxx" here which is not supported in aws built-in configuration file.